src/layout/cursor.rs

```diff
@@ -5,15 +5,24 @@ use super::*;
 /// Represents a position within a layout.
 #[derive(Copy, Clone, Default, Debug)]
 pub struct Cursor {
-    path: CursorPath,
-    baseline: f32,
-    offset: f32,
-    advance: f32,
-    text_start: usize,
-    text_end: usize,
-    is_rtl: bool,
-    is_leading: bool,
-    is_inside: bool,
+    /// path to the target cluster
+    pub path: CursorPath,
+    /// offset to the baseline
+    pub baseline: f32,
+    /// offset to the target cluster along the baseline
+    pub offset: f32,
+    /// advance of the target cluster
+    pub advance: f32,
+    /// start of the target cluster
+    pub text_start: usize,
```

Comment:
I’m torn on whether to make these fields an actual `Range<usize>`. I think it makes sense now, but I have some planned changes that will make these types a bit more complicated to handle text that spans multiple source objects and `Range` may not fit that. 

I think we’ll leave it as is for now.

Reply:
`Range` is annoying because it doesn't do `Copy` and breaks the derive macro on this struct. Bare fields are more useful, and you prettymuch always need to reconstruct ranges anyway to use them in more than one place.